# SETUP-System-For-Colab


## Download Files from WEB

In [3]:
!git clone https://github.com/cocodataset/cocoapi.git  
maketext="""
all:
    # install pycocotools locally
	python3 setup.py build_ext --inplace
	rm -rf build

install:
	# install pycocotools to the Python site-packages
	python3 setup.py build_ext install
	rm -rf build
"""
print(maketext, file=open('/content/cocoapi/PythonAPI/Makefile', 'w'))
setuptxt="""
from setuptools import setup, Extension
import numpy as np

# To compile and install locally run "python setup.py build_ext --inplace"
# To install library to Python site-packages run "python setup.py build_ext install"

ext_modules = [
    Extension(
        'pycocotools._mask',
        sources=['../common/maskApi.c', 'pycocotools/_mask.pyx'],
        #sources=['common/maskApi.c'],
        include_dirs = [np.get_include(), '../common'],
        #extra_compile_args=['-Wno-cpp', '-Wno-unused-function', '-std=c99'],
        extra_compile_args=['-std=c99'],
    )
]

setup(
    name='pycocotools',
    packages=['pycocotools'],
    package_dir = {'pycocotools': 'pycocotools'},
    install_requires=[
        'setuptools>=18.0',
        'cython>=0.27.3',
        'matplotlib>=2.1.0'
    ],
    version='2.0',
    ext_modules= ext_modules
)
"""
print(setuptxt, file=open('/content/cocoapi/PythonAPI/setup.py', 'w'))
!cd /content/cocoapi/PythonAPI; make; make install
!mkdir /content/cocoapi/images/
!mkdir /content/cocoapi/annotations/
!mkdir models
!wget http://images.cocodataset.org/zips/train2014.zip
!wget http://images.cocodataset.org/zips/val2014.zip
!wget http://images.cocodataset.org/zips/test2014.zip
!wget http://images.cocodataset.org/annotations/annotations_trainval2014.zip
!wget http://images.cocodataset.org/annotations/image_info_test2014.zip
!unzip /content/train2014.zip -d /content/cocoapi/images/
!unzip /content/test2014.zip  -d /content/cocoapi/images/
!unzip /content/val2014.zip  -d /content/cocoapi/images/ 
!unzip /content/annotations_trainval2014.zip  -d /content/cocoapi/annotations/
!unzip /content/image_info_test2014.zip -d /content/cocoapi/annotations/
!mv /content/cocoapi/annotations/annotations/* /content/cocoapi/annotations/
!rm -rf /content/cocoapi/annotations/annotations
!rm /content/train2014.zip
!rm /content/test2014.zip
!rm /content/val2014.zip
!rm /content/annotations_trainval2014.zip
!rm /content/image_info_test2014.zip
!pip install nltk
import nltk
nltk.download('punkt')

Streaming output truncated to the last 5000 lines.
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000029431.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000036349.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000502055.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000438848.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000364973.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000564166.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000323327.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000338581.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000039656.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000071667.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000376046.jpg  
 extracting: /content/cocoapi/images/val2014/COCO_val2014_000000066394.jpg  
 extracting: /content/coc

True

## Copy Files from GDRIVE

In [1]:
from google.colab import drive
drive.mount('/content/drive')
!git clone https://github.com/cocodataset/cocoapi.git  
maketext="""
all:
    # install pycocotools locally
	python3 setup.py build_ext --inplace
	rm -rf build

install:
	# install pycocotools to the Python site-packages
	python3 setup.py build_ext install
	rm -rf build
"""
print(maketext, file=open('/content/cocoapi/PythonAPI/Makefile', 'w'))
setuptxt="""
from setuptools import setup, Extension
import numpy as np

# To compile and install locally run "python setup.py build_ext --inplace"
# To install library to Python site-packages run "python setup.py build_ext install"

ext_modules = [
    Extension(
        'pycocotools._mask',
        sources=['../common/maskApi.c', 'pycocotools/_mask.pyx'],
        #sources=['common/maskApi.c'],
        include_dirs = [np.get_include(), '../common'],
        #extra_compile_args=['-Wno-cpp', '-Wno-unused-function', '-std=c99'],
        extra_compile_args=['-std=c99'],
    )
]

setup(
    name='pycocotools',
    packages=['pycocotools'],
    package_dir = {'pycocotools': 'pycocotools'},
    install_requires=[
        'setuptools>=18.0',
        'cython>=0.27.3',
        'matplotlib>=2.1.0'
    ],
    version='2.0',
    ext_modules= ext_modules
)
"""
print(setuptxt, file=open('/content/cocoapi/PythonAPI/setup.py', 'w'))
!cd /content/cocoapi/PythonAPI; make; make install
!mkdir /content/cocoapi/images/
!mkdir /content/cocoapi/annotations/
!mkdir models
!cp /content/drive/MyDrive/_Colab/CV-Image_Captioning/*.py /content/
!cp /content/drive/MyDrive/_Colab/CV-Image_Captioning/cocoapi/images/*.zip /content/
!cp /content/drive/MyDrive/_Colab/CV-Image_Captioning/cocoapi/annotations/*.zip /content/
!cp /content/drive/MyDrive/_Colab/CV-Image_Captioning/models/* /content/models/
!unzip /content/train2014.zip -d /content/cocoapi/images/
!unzip /content/test2014.zip  -d /content/cocoapi/images/
!unzip /content/val2014.zip  -d /content/cocoapi/images/ 
!unzip /content/annotations_trainval2014.zip  -d /content/cocoapi/annotations/
!unzip /content/image_info_test2014.zip -d /content/cocoapi/annotations/
!mv /content/cocoapi/annotations/annotations/* /content/cocoapi/annotations/
!rm -rf /content/cocoapi/annotations/annotations
!rm /content/train2014.zip
!rm /content/test2014.zip
!rm /content/val2014.zip
!rm /content/annotations_trainval2014.zip
!rm /content/image_info_test2014.zip
!pip install nltk
import nltk
nltk.download('punkt')

Mounted at /content/drive
Cloning into 'cocoapi'...
remote: Enumerating objects: 975, done.
remote: Total 975 (delta 0), reused 0 (delta 0), pack-reused 975
Receiving objects: 100% (975/975), 11.72 MiB | 11.89 MiB/s, done.
Resolving deltas: 100% (576/576), done.
python3 setup.py build_ext --inplace
running build_ext
cythoning pycocotools/_mask.pyx to pycocotools/_mask.c
/usr/local/lib/python3.7/dist-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /content/cocoapi/PythonAPI/pycocotools/_mask.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)
building 'pycocotools._mask' extension
creating build
creating build/common
creating build/temp.linux-x86_64-3.7
creating build/temp.linux-x86_64-3.7/pycocotools
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fdebug-prefix-map=/build/python3.7-OGiuun/python3.7-3.7.10=. -fstack-prote

True

## Get from Gdrive Small DB

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
!git clone https://github.com/cocodataset/cocoapi.git  
maketext="""
all:
    # install pycocotools locally
	python3 setup.py build_ext --inplace
	rm -rf build

install:
	# install pycocotools to the Python site-packages
	python3 setup.py build_ext install
	rm -rf build
"""
print(maketext, file=open('/content/cocoapi/PythonAPI/Makefile', 'w'))
setuptxt="""
from setuptools import setup, Extension
import numpy as np

# To compile and install locally run "python setup.py build_ext --inplace"
# To install library to Python site-packages run "python setup.py build_ext install"

ext_modules = [
    Extension(
        'pycocotools._mask',
        sources=['../common/maskApi.c', 'pycocotools/_mask.pyx'],
        #sources=['common/maskApi.c'],
        include_dirs = [np.get_include(), '../common'],
        #extra_compile_args=['-Wno-cpp', '-Wno-unused-function', '-std=c99'],
        extra_compile_args=['-std=c99'],
    )
]

setup(
    name='pycocotools',
    packages=['pycocotools'],
    package_dir = {'pycocotools': 'pycocotools'},
    install_requires=[
        'setuptools>=18.0',
        'cython>=0.27.3',
        'matplotlib>=2.1.0'
    ],
    version='2.0',
    ext_modules= ext_modules
)
"""
print(setuptxt, file=open('/content/cocoapi/PythonAPI/setup.py', 'w'))
!cd /content/cocoapi/PythonAPI; make; make install
!mkdir /content/cocoapi/images/
!mkdir /content/cocoapi/annotations/
!mkdir models
!cp /content/drive/MyDrive/_Colab/CV-Image_Captioning/*.py /content/
!cp /content/drive/MyDrive/_Colab/*_small.zip /content/
!cp /content/drive/MyDrive/_Colab/CV-Image_Captioning/models/* /content/models/
!unzip /content/train2014_small.zip -d /content/cocoapi/images/
!unzip /content/test2014_small.zip  -d /content/cocoapi/images/
!unzip /content/val2014_small.zip  -d /content/cocoapi/images/ 
!unzip /content/annotations_trainval2014_small.zip  -d /content/cocoapi/annotations/
!unzip /content/annotations_small.zip -d /content/cocoapi/annotations/
!mv /content/cocoapi/annotations/annotations/* /content/cocoapi/annotations/
!rm -rf /content/cocoapi/annotations/annotations
!rm /content/train2014_small.zip
!rm /content/test2014_small.zip
!rm /content/val2014_small.zip
!rm /content/annotations_small.zip
!pip install nltk
import nltk
nltk.download('punkt')

KeyboardInterrupt: ignored


# Computer Vision Nanodegree

## Project: Image Captioning

---

In this notebook, you will train your CNN-RNN model.  

You are welcome and encouraged to try out many different architectures and hyperparameters when searching for a good model.

This does have the potential to make the project quite messy!  Before submitting your project, make sure that you clean up:
- the code you write in this notebook.  The notebook should describe how to train a single CNN-RNN architecture, corresponding to your final choice of hyperparameters.  You should structure the notebook so that the reviewer can replicate your results by running the code in this notebook.  
- the output of the code cell in **Step 2**.  The output should show the output obtained when training the model from scratch.

This notebook **will be graded**.  

Feel free to use the links below to navigate the notebook:
- [Step 1](#step1): Training Setup
- [Step 2](#step2): Train your Model
- [Step 3](#step3): (Optional) Validate your Model

<a id='step1'></a>
## Step 1: Training Setup

In this step of the notebook, you will customize the training of your CNN-RNN model by specifying hyperparameters and setting other options that are important to the training procedure.  The values you set now will be used when training your model in **Step 2** below.

You should only amend blocks of code that are preceded by a `TODO` statement.  **Any code blocks that are not preceded by a `TODO` statement should not be modified**.

### Task #1

Begin by setting the following variables:
- `batch_size` - the batch size of each training batch.  It is the number of image-caption pairs used to amend the model weights in each training step. 
- `vocab_threshold` - the minimum word count threshold.  Note that a larger threshold will result in a smaller vocabulary, whereas a smaller threshold will include rarer words and result in a larger vocabulary.  
- `vocab_from_file` - a Boolean that decides whether to load the vocabulary from file. 
- `embed_size` - the dimensionality of the image and word embeddings.  
- `hidden_size` - the number of features in the hidden state of the RNN decoder.  
- `num_epochs` - the number of epochs to train the model.  We recommend that you set `num_epochs=3`, but feel free to increase or decrease this number as you wish.  [This paper](https://arxiv.org/pdf/1502.03044.pdf) trained a captioning model on a single state-of-the-art GPU for 3 days, but you'll soon see that you can get reasonable results in a matter of a few hours!  (_But of course, if you want your model to compete with current research, you will have to train for much longer._)
- `save_every` - determines how often to save the model weights.  We recommend that you set `save_every=1`, to save the model weights after each epoch.  This way, after the `i`th epoch, the encoder and decoder weights will be saved in the `models/` folder as `encoder-i.pkl` and `decoder-i.pkl`, respectively.
- `print_every` - determines how often to print the batch loss to the Jupyter notebook while training.  Note that you **will not** observe a monotonic decrease in the loss function while training - this is perfectly fine and completely expected!  You are encouraged to keep this at its default value of `100` to avoid clogging the notebook, but feel free to change it.
- `log_file` - the name of the text file containing - for every step - how the loss and perplexity evolved during training.

If you're not sure where to begin to set some of the values above, you can peruse [this paper](https://arxiv.org/pdf/1502.03044.pdf) and [this paper](https://arxiv.org/pdf/1411.4555.pdf) for useful guidance!  **To avoid spending too long on this notebook**, you are encouraged to consult these suggested research papers to obtain a strong initial guess for which hyperparameters are likely to work best.  Then, train a single model, and proceed to the next notebook (**3_Inference.ipynb**).  If you are unhappy with your performance, you can return to this notebook to tweak the hyperparameters (and/or the architecture in **model.py**) and re-train your model.

### Question 1

**Question:** Describe your CNN-RNN architecture in detail.  With this architecture in mind, how did you select the values of the variables in Task 1?  If you consulted a research paper detailing a successful implementation of an image captioning model, please provide the reference.

**Answer:** 
* Started with a simple DecoderRNN of: 1 Embedding Layer -> LSTM -> Linear
* Tried to add an attention layer after LSTM: 1 Embedding Layer -> LSTM -> MultiHeadAttention -> Linear
* Tried to add an attention layer before LSTM: 1 Embedding Layer -> MultiHeadAttention -> LSTM -> Linear
* Found in the internet literature on best case performances and how they did it. Particularly found [GitHub Example](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning) as well as [this medium tutorial](https://medium.com/analytics-vidhya/image-captioning-with-attention-part-1-e8a5f783f6d3)
* Compared performance of all models with an excelent example from: [here](https://medium.com/analytics-vidhya/image-captioning-with-attention-part-1-e8a5f783f6d3)
Then realised that my best model was the simplest one v1.0.2 close in performance with the more complex one used for benchmark found at  [this medium tutorial](https://medium.com/analytics-vidhya/image-captioning-with-attention-part-1-e8a5f783f6d3)


### (Optional) Task #2

Note that we have provided a recommended image transform `transform_train` for pre-processing the training images, but you are welcome (and encouraged!) to modify it as you wish.  When modifying this transform, keep in mind that:
- the images in the dataset have varying heights and widths, and 
- if using a pre-trained model, you must perform the corresponding appropriate normalization.

### Question 2

**Question:** How did you select the transform in `transform_train`?  If you left the transform at its provided value, why do you think that it is a good choice for your CNN architecture?

**Answer:** I removed the Flip transformation. I believe it was slowing down and performance wise didn't see much added value given that most pictures were aligned. In terms of performance if enhanced the model. I would consider further testing in case I see any issues due to this "simplification".

### Task #3

Next, you will specify a Python list containing the learnable parameters of the model.  For instance, if you decide to make all weights in the decoder trainable, but only want to train the weights in the embedding layer of the encoder, then you should set `params` to something like:
```
params = list(decoder.parameters()) + list(encoder.embed.parameters()) 
```

### Question 3

**Question:** How did you select the trainable parameters of your architecture?  Why do you think this is a good choice?

**Answer:** Removed the encoder parameters as these are already pretrained, didn't see much added in having them at the optimizer. Believe having them could add to problems in performance and added noise.

### Task #4

Finally, you will select an [optimizer](http://pytorch.org/docs/master/optim.html#torch.optim.Optimizer).

### Question 4

**Question:** How did you select the optimizer used to train your model?

**Answer:** Selected Adam, as per my intuition, experience, literature learned, ... performs better for this gradient questions, due to saving previous stages. As suggested in my previous project, one can take a look at this [comparisation Adam vs SGD](https://medium.com/@Biboswan98/optim-adam-vs-optim-sgd-lets-dive-in-8dbf1890fbdc). However I did lower the learning rate to 0.0001.

In [4]:
import torch
import torch.nn as nn
from torchvision import transforms
import sys
sys.path.append('cocoapi/PythonAPI')
from pycocotools.coco import COCO
from data_loader import get_loader
from model import EncoderCNN, DecoderRNNv102, DecoderRNNv120
import math
 
 
## TODO #1: Select appropriate values for the Python variables below.
batch_size = 10          # batch size
vocab_threshold = 3        # minimum word count threshold
vocab_from_file = False    # if True, load existing vocab file
embed_size = 512           # dimensionality of image and word embeddings
hidden_size = 1024          # number of features in hidden state of the RNN decoder
num_epochs = 10             # number of training epochs
save_every = 1             # determines frequency of saving model weights
print_every = 1000          # determines window for printing average loss
log_file = 'training_log.txt'       # name of file with saved training loss and perplexity
VERSION = 'v1.0.2'
SUBVERSION = 'params:Decoder_withoutFlipTrans'
_TMP = [VERSION, SUBVERSION, f'batch_size:{batch_size}', f'vocabThr:{vocab_threshold}', f'embedSize:{embed_size}',
        f'hiddenSize:{hidden_size}',f'totEpochs:{num_epochs}']
RUN_KEY = "_".join(_TMP)
FILENAME = "".join([c for c in RUN_KEY if c not in '( ,:).'])
 
# My variables
USE_REQUEST = False
 
# (Optional) TODO #2: Amend the image transform below.
transform_train = transforms.Compose([ 
    transforms.Resize(256),                          # smaller edge of image resized to 256
    transforms.RandomCrop(224),                      # get 224x224 crop from random location
    #transforms.RandomHorizontalFlip(0.5),               # horizontally flip image with probability=0.5
    transforms.ToTensor(),                           # convert the PIL Image to a tensor
    transforms.Normalize((0.485, 0.456, 0.406),      # normalize image for pre-trained model
                         (0.229, 0.224, 0.225))])
 
 
# Build data loader.
data_loader = get_loader(transform=transform_train,
                         mode='train',
                         batch_size=batch_size,
                         vocab_threshold=vocab_threshold,
                         vocab_from_file=vocab_from_file, cocoapi_loc='')
 
# The size of the vocabulary.
vocab_size = len(data_loader.dataset.vocab)
 
# Initialize the encoder and decoder. 
encoder = EncoderCNN(embed_size)
decoder = DecoderRNNv102(embed_size, hidden_size, vocab_size)
 
 
# Move models to GPU if CUDA is available. 
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
encoder.to(device)
decoder.to(device)
 
# Define the loss function. 
criterion = nn.CrossEntropyLoss().cuda() if torch.cuda.is_available() else nn.CrossEntropyLoss()
 
# TODO #3: Specify the learnable parameters of the model.
params = decoder.parameters()
 
# TODO #4: Define the optimizer.
optimizer = torch.optim.Adam(params, lr=0.0001, betas=(0.9, 0.999))
 
# Set the total number of training steps per epoch.
total_step = math.ceil(len(data_loader.dataset.caption_lengths) / data_loader.batch_sampler.batch_size)

loading annotations into memory...
Done (t=0.88s)
creating index...
index created!
[0/414113] Tokenizing captions...
[100000/414113] Tokenizing captions...
[200000/414113] Tokenizing captions...
[300000/414113] Tokenizing captions...
[400000/414113] Tokenizing captions...
loading annotations into memory...
Done (t=0.65s)
creating index...


  0%|          | 884/414113 [00:00<00:46, 8837.21it/s]

index created!
Obtaining caption lengths...


100%|██████████| 414113/414113 [00:42<00:00, 9631.37it/s]
Downloading: "https://download.pytorch.org/models/resnet50-19c8e357.pth" to /root/.cache/torch/hub/checkpoints/resnet50-19c8e357.pth


<a id='step2'></a>
## Step 2: Train your Model

Once you have executed the code cell in **Step 1**, the training procedure below should run without issue.  

It is completely fine to leave the code cell below as-is without modifications to train your model.  However, if you would like to modify the code used to train the model below, you must ensure that your changes are easily parsed by your reviewer.  In other words, make sure to provide appropriate comments to describe how your code works!  

You may find it useful to load saved weights to resume training.  In that case, note the names of the files containing the encoder and decoder weights that you'd like to load (`encoder_file` and `decoder_file`).  Then you can load the weights by using the lines below:

```python
# Load pre-trained weights before resuming training.
encoder.load_state_dict(torch.load(os.path.join('./models', encoder_file)))
decoder.load_state_dict(torch.load(os.path.join('./models', decoder_file)))
```

While trying out parameters, make sure to take extensive notes and record the settings that you used in your various training runs.  In particular, you don't want to encounter a situation where you've trained a model for several hours but can't remember what settings you used :).

### A Note on Tuning Hyperparameters

To figure out how well your model is doing, you can look at how the training loss and perplexity evolve during training - and for the purposes of this project, you are encouraged to amend the hyperparameters based on this information.  

However, this will not tell you if your model is overfitting to the training data, and, unfortunately, overfitting is a problem that is commonly encountered when training image captioning models.  

For this project, you need not worry about overfitting. **This project does not have strict requirements regarding the performance of your model**, and you just need to demonstrate that your model has learned **_something_** when you generate captions on the test data.  For now, we strongly encourage you to train your model for the suggested 3 epochs without worrying about performance; then, you should immediately transition to the next notebook in the sequence (**3_Inference.ipynb**) to see how your model performs on the test data.  If your model needs to be changed, you can come back to this notebook, amend hyperparameters (if necessary), and re-train the model.

That said, if you would like to go above and beyond in this project, you can read about some approaches to minimizing overfitting in section 4.3.1 of [this paper](http://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=7505636).  In the next (optional) step of this notebook, we provide some guidance for assessing the performance on the validation dataset.

In [10]:
import os
encoder_file = 'v102_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs10encoder-5.pkl'
decoder_file = 'v102_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs10decoder-5.pkl'
encoder.load_state_dict(torch.load(os.path.join('./models', encoder_file)))
decoder.load_state_dict(torch.load(os.path.join('./models', decoder_file)))

<All keys matched successfully>

In [9]:
import torch.utils.data as data
import numpy as np
import os
import requests
import time
 
# Open the training log file.
f = open(FILENAME + log_file, 'a')
print(f'RUN_KEY: {RUN_KEY}', file = f)
 
 
old_time = time.time()
if USE_REQUEST:
    response = requests.request("GET", 
                                "http://metadata.google.internal/computeMetadata/v1/instance/attributes/keep_alive_token", 
                                headers={"Metadata-Flavor":"Google"})
 
loss_arr, perplexity = [], []
 
for epoch in range(1, num_epochs+1):
    
    for i_step in range(1, total_step+1):
        encoder.eval() # no fine-tuning for Encoder
        decoder.train()
          
        # Randomly sample a caption length, and sample indices with that length.
        indices = data_loader.dataset.get_train_indices()
        # Create and assign a batch sampler to retrieve a batch with the sampled indices.
        new_sampler = data.sampler.SubsetRandomSampler(indices=indices)
        data_loader.batch_sampler.sampler = new_sampler
        
        # Obtain the batch.
        images, captions = next(iter(data_loader))
        
        # Move batch of images and captions to GPU if CUDA is available.
        images = images.to(device)
        captions = captions.to(device)
         # target captions, excluding the first word
        captions_target = captions[:, 1:]
        # captions for training without the last word
        captions_train = captions[:, :-1]
        
        # Zero the gradients.
        decoder.zero_grad()
        encoder.zero_grad()
        
        # Pass the inputs through the CNN-RNN model.
        features = encoder(images)
        outputs = decoder(captions= captions_train, features = features)
        
        # Calculate the batch loss.
        loss = criterion(outputs.view(-1, vocab_size), captions.view(-1))
        
        # Backward pass.
        loss.backward()
        
        # Update the parameters in the optimizer.
        optimizer.step()
            
        # Get training statistics.
        
        loss_arr.append(loss.item())
        perplexity.append(np.exp(loss.item()))
        # Print training statistics (on same line).
        
        if i_step % (print_every // 10) == 0:
            dur = time.time() - old_time
            old_time = time.time()
            stats = f'[{dur:.2f}s] Epoch [{epoch}/{num_epochs}], Step [{i_step}/{total_step}], Loss: {loss.item():.4f}, Perplexity: {np.exp(loss.item()):5.4f}'
            if i_step % print_every == 0: 
              print('\r' + stats) # Print training statistics (on different line).
              # Print training statistics to file.
              print(stats + '\n', file=f)
        
              if USE_REQUEST:
                  requests.request("POST",  "https://nebula.udacity.com/api/v1/remote/keep-alive",  headers={'Authorization': "STAR " + response.text})
        
            else: print('\r' + stats, end="")
            
    # Save the weights.
    if epoch % save_every == 0:
 
        torch.save(decoder.state_dict(), os.path.join('./models', FILENAME + 'decoder-%d.pkl' % epoch))
        torch.save(encoder.state_dict(), os.path.join('./models', FILENAME + 'encoder-%d.pkl' % epoch))
        np.save(open(f'{FILENAME}-loss_arr.np', 'wb'), np.array(loss_arr))
        np.save(open(f'{FILENAME}-perplexity.np', 'wb'), np.array(perplexity))
        !cp /content/models/* /content/drive/MyDrive/_Colab/CV-Image_Captioning/models/
        !cp /content/*txt /content/drive/MyDrive/_Colab/CV-Image_Captioning/
        !cp /content/*np /content/drive/MyDrive/_Colab/CV-Image_Captioning/
 
# Close the training log file.
#f.close()
print('\n', min(loss_arr), min(perplexity))


[35.50s] Epoch [3/10], Step [1000/41412], Loss: 2.3687, Perplexity: 10.6840
[34.17s] Epoch [3/10], Step [2000/41412], Loss: 2.4524, Perplexity: 11.6158
[33.98s] Epoch [3/10], Step [3000/41412], Loss: 1.9602, Perplexity: 7.1007
[34.30s] Epoch [3/10], Step [4000/41412], Loss: 2.2154, Perplexity: 9.1647
[33.62s] Epoch [3/10], Step [5000/41412], Loss: 1.8282, Perplexity: 6.2229
[33.30s] Epoch [3/10], Step [6000/41412], Loss: 2.1126, Perplexity: 8.2698
[33.25s] Epoch [3/10], Step [7000/41412], Loss: 2.3057, Perplexity: 10.0316
[33.16s] Epoch [3/10], Step [8000/41412], Loss: 1.6720, Perplexity: 5.3229
[33.40s] Epoch [3/10], Step [9000/41412], Loss: 2.2362, Perplexity: 9.3578
[33.03s] Epoch [3/10], Step [10000/41412], Loss: 1.6791, Perplexity: 5.3606
[33.28s] Epoch [3/10], Step [11000/41412], Loss: 1.9335, Perplexity: 6.9140
[33.01s] Epoch [3/10], Step [12000/41412], Loss: 2.0442, Perplexity: 7.7233
[33.08s] Epoch [3/10], Step [13000/41412], Loss: 2.0054, Perplexity: 7.4293
[33.16s] Epoch [3

<a id='step3'></a>
## Step 3: (Optional) Validate your Model

To assess potential overfitting, one approach is to assess performance on a validation set.  If you decide to do this **optional** task, you are required to first complete all of the steps in the next notebook in the sequence (**3_Inference.ipynb**); as part of that notebook, you will write and test code (specifically, the `sample` method in the `DecoderRNN` class) that uses your RNN decoder to generate captions.  That code will prove incredibly useful here. 

If you decide to validate your model, please do not edit the data loader in **data_loader.py**.  Instead, create a new file named **data_loader_val.py** containing the code for obtaining the data loader for the validation data.  You can access:
- the validation images at filepath `'/opt/cocoapi/images/train2014/'`, and
- the validation image caption annotation file at filepath `'/opt/cocoapi/annotations/captions_val2014.json'`.

The suggested approach to validating your model involves creating a json file such as [this one](https://github.com/cocodataset/cocoapi/blob/master/results/captions_val2014_fakecap_results.json) containing your model's predicted captions for the validation images.  Then, you can write your own script or use one that you [find online](https://github.com/tylin/coco-caption) to calculate the BLEU score of your model.  You can read more about the BLEU score, along with other evaluation metrics (such as TEOR and Cider) in section 4.1 of [this paper](https://arxiv.org/pdf/1411.4555.pdf).  For more information about how to use the annotation file, check out the [website](http://cocodataset.org/#download) for the COCO dataset.

## Auxiliary Eval Func

### Loading models based on Run-Key and epoch

In [11]:
def loadTrainedModels(run, epoch, dictname):
    from model import EncoderCNNv1, EncoderCNN
    from model import DecoderRNNv101, DecoderRNNv102, DecoderRNNv120, DecoderRNNv121, DecoderRNNv200
    import pickle
    import torch
    import os
    from vocabulary import Vocabulary
    import torch

    encoder_file = run + f'encoder-{epoch}.pkl'
    decoder_file = run + f'decoder-{epoch}.pkl'
    batchsize = int(run[run.find('_batch_size') + 11:run.find('_vocabThr')])
    embed_size = int(run[run.find('embedSize') + 9:run.find('_hiddenSize')])
    hidden_size = int(run[run.find('_hiddenSize') + 11:run.find('_totEpochs')])
    smallDB = True if 'small' in run else False
    
    voca = pickle.load(open(dictname, 'rb'))
    
    vocab_size = len(voca)
    #vocab_size = 11498 # 9955 #5892

    encoder = EncoderCNN(embed_size)
    model = run[:run.find('_')]
    if model == 'v101' or model == 'v100':  decoder = DecoderRNNv101(embed_size, hidden_size, vocab_size)
    elif model == 'v102':  decoder = DecoderRNNv102(embed_size, hidden_size, vocab_size)
    elif model == 'v110':  decoder = DecoderRNNv110(embed_size, hidden_size, vocab_size)
    elif model == 'v120':  decoder = DecoderRNNv120(embed_size, hidden_size, vocab_size)
    elif model == 'v121':  decoder = DecoderRNNv121(embed_size, hidden_size, vocab_size)
    elif model == 'v200':  decoder = DecoderRNNv200(2048, embed_size, hidden_size, vocab_size)

    # Move models to GPU if CUDA is available. 
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    encoder.to(device)
    decoder.to(device)
    print(run)
    encoder.load_state_dict(torch.load(os.path.join('./models', encoder_file), map_location=device) )
    decoder.load_state_dict(torch.load(os.path.join('./models', decoder_file), map_location=device))

    return encoder, decoder


### Clean Sentence

In [12]:
def clean_sentence(data_loader, output):
    sentence = [data_loader.dataset.vocab.idx2word[f] for f in output]
    error = False
    if sentence[0] != data_loader.dataset.vocab.start_word or sentence[-1] != data_loader.dataset.vocab.end_word: 
      error = True
    if sentence[0] == data_loader.dataset.vocab.start_word: sentence = sentence[1:]
    if sentence[-1] == data_loader.dataset.vocab.end_word: sentence = sentence[:-1]
    #end of line
    if sentence == None: return None, True
    if len(sentence) > 1 and sentence[-1] in '!.?...…':  sentence=" ".join(sentence[:-1]) + sentence[-1]
    else: sentence=" ".join(sentence)
    return sentence, error

### Return Predicted Dictionary

In [13]:
def generateValDict(encoder, decoder, vocab_file='vocab_complete_training_thr4.pkl', folder_vocab='/content/'):
    import numpy as np
    import matplotlib.pyplot as plt
    from torchvision import transforms
    from data_loader_val import get_val_loader
    import torch
    %matplotlib inline

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")    

    transform_test = transforms.Compose([ 
        transforms.Resize(256),                          # smaller edge of image resized to 256
        transforms.RandomCrop(224),                      # get 224x224 crop from random location
        #transforms.RandomHorizontalFlip(0.5),               # horizontally flip image with probability=0.5
        transforms.ToTensor(),                           # convert the PIL Image to a tensor
        transforms.Normalize((0.485, 0.456, 0.406),      # normalize image for pre-trained model
                            (0.229, 0.224, 0.225))])


    # Create the data loader.
    vocab_file = folder_vocab + vocab_file
    data_loader_val = get_val_loader(transform=transform_test, vocab_file=vocab_file, 
                                     vocab_from_file=True ,mode='val', cocoapi_loc='')
    result = []
    prob_ids = []
    for i, (idx, orig_image, image, caption) in enumerate(data_loader_val):
        print(f'\rPredicting {i:<5} of {len(data_loader_val)}', end='')
        image = image.to(device)
        features = encoder(image).unsqueeze(1)
        output = decoder.sample(features)
        
        a = dict()
        a['image_id'] = idx.item()
        cap, err = clean_sentence(data_loader_val, output)
        a['caption'] = cap
        #print(f" id: {a['image_id']}: ({a['caption']}) - Real {caption}")
        if cap != None: result.append(a)
        if err: prob_ids.append(a['image_id'])
        #if i == 20: return
    
    return result, prob_ids

### Load Evaluation Engine and Evaluate

In [14]:
!git clone https://github.com/mtanti/coco-caption.git
!sh /content/coco-caption/get_stanford_models.sh
!sh /content/coco-caption/get_google_word2vec_model.sh
# in case cookie doesn't work on get_google_word2vec_model.sh script
#!wget -c "https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz"
#!mv /content/GoogleNews-vectors-negative300.bin.gz /content/coco-caption/pycocoevalcap/wmd/data/
#!gzip -d /content/coco-caption/pycocoevalcap/wmd/data/GoogleNews-vectors-negative300.bin.gz

def evaluateRun(run_key):
    %matplotlib inline
    from pycocotools.coco import COCO
    import sys
    sys.path.append('coco-caption')
    sys.path.append('coco-caption/pycocoevalcap')
    sys.path.append('coco-caption/pycocoevalcap/bleu')
    sys.path.append('coco-caption/pycocoevalcap/cider')
    sys.path.append('coco-caption/pycocoevalcap/meteor')
    sys.path.append('coco-caption/pycocoevalcap/rouge')
    sys.path.append('coco-caption/pycocoevalcap/spice')
    sys.path.append('coco-caption/pycocoevalcap/tokenizer')
    from pycocoevalcap.eval import COCOEvalCap
    import matplotlib.pyplot as plt
    import skimage.io as io
    import pylab
    pylab.rcParams['figure.figsize'] = (10.0, 8.0)

    import json
    from json import encoder
    encoder.FLOAT_REPR = lambda o: format(o, '.3f')

    # set up file names and pathes
    dataDir='/content/cocoapi'
    dataType='val2014'
    algName = run_key
    annFile=f'{dataDir}/annotations/captions_{dataType}.json'
    subtypes=['results', 'evalImgs', 'eval']
    [resFile, evalImgsFile, evalFile]= \
    [f'{dataDir}/results/captions_{dataType}_{algName}_{subtype}.json' for subtype in subtypes]

    # create coco object and cocoRes object
    coco = COCO(annFile)
    cocoRes = coco.loadRes(resFile)

    # create cocoEval object by taking coco and cocoRes
    cocoEval = COCOEvalCap(coco, cocoRes)

    # evaluate on a subset of images by setting
    # cocoEval.params['image_id'] = cocoRes.getImgIds()
    # please remove this line when evaluating the full validation set
    cocoEval.params['image_id'] = cocoRes.getImgIds()

    # evaluate results
    # SPICE will take a few minutes the first time, but speeds up due to caching
    cocoEval.evaluate()

    # print output evaluation scores
    for metric, score in cocoEval.eval.items():
        print( f'{metric}: {score}')

    # save evaluation results to ./results folder
    json.dump(cocoEval.evalImgs, open(evalImgsFile, 'w'))
    json.dump(cocoEval.eval,     open(evalFile, 'w'))

    # plot score histogram
    ciderScores = [eva['CIDEr'] for eva in cocoEval.evalImgs]
    plt.hist(ciderScores)
    plt.title('Histogram of CIDEr Scores', fontsize=20)
    plt.xlabel('CIDEr score', fontsize=20)
    plt.ylabel('result counts', fontsize=20)
    plt.show()

    return cocoEval.eval

Cloning into 'coco-caption'...
remote: Enumerating objects: 763, done.
remote: Total 763 (delta 0), reused 0 (delta 0), pack-reused 763
Receiving objects: 100% (763/763), 130.05 MiB | 15.56 MiB/s, done.
Resolving deltas: 100% (398/398), done.
Downloading...
--2021-06-17 15:15:48--  http://nlp.stanford.edu/software/stanford-corenlp-full-2015-12-09.zip
Resolving nlp.stanford.edu (nlp.stanford.edu)... 171.64.67.140
Connecting to nlp.stanford.edu (nlp.stanford.edu)|171.64.67.140|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://nlp.stanford.edu/software/stanford-corenlp-full-2015-12-09.zip [following]
--2021-06-17 15:15:48--  https://nlp.stanford.edu/software/stanford-corenlp-full-2015-12-09.zip
Connecting to nlp.stanford.edu (nlp.stanford.edu)|171.64.67.140|:443... connected.
HTTP request sent, awaiting response... 302 FOUND
Location: https://downloads.cs.stanford.edu/nlp/software/stanford-corenlp-full-2015-12-09.zip [following]
--2021-06-17 15:15:49-- 

## Evaluate Model

In [ ]:
# (Optional) TODO: Validate your model.
import json

#run_key = 'v200_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr4_embedSize512_hiddenSize1024_totEpochs10'
#run_key = 'v102_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs10'
#run_key = 'v121_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs3'
#run_key = 'v120_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs3'

run_key = 'v102_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs10'

epoch = 5

encoder, decoder  = loadTrainedModels(run_key, epoch, 'vocab.pkl')
encoder.eval()
decoder.eval()

results, wrongIds = generateValDict(encoder, decoder, 'vocab.pkl')
json.dump(results, open(f"/content/cocoapi/results/captions_val2014_{run_key}_results.json", "w"))
Predicting 40503 of 40504""")
evaluateRun(run_key)


v102_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs10
Vocabulary successfully loaded from vocab.pkl file!
Predicting 40503 of 40504
loading annotations into memory...
Done (t=0.74s)
creating index...
index created!
Loading and preparing results...
DONE (t=0.10s)
creating index...
index created!
tokenization...
setting up scorers...
computing Bleu score...
{'testlen': 396775, 'reflen': 391861, 'guess': [396775, 356271, 315767, 275263], 'correct': [250627, 112143, 45373, 18988]}
ratio: 1.012540160924407
Bleu_1: 0.632
Bleu_2: 0.446
Bleu_3: 0.306
Bleu_4: 0.211
computing METEOR score...
METEOR: 0.213
computing Rouge score...
ROUGE_L: 0.466
computing CIDEr score...
CIDEr: 0.668
computing SPICE score...


## Results

**Benchmark** (`v200_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr4_embedSize512_hiddenSize1024_totEpochs10`)
* Attention: 2 epochs: 
> ratio: 0.9943724833842694; Bleu_1: 0.679; Bleu_2: 0.496; Bleu_3: 0.344; Bleu_4: 0.232; METEOR: 0.211; ROUGE_L: 0.491; CIDEr: 0.639
* Attention: 3 epochs: 
> ratio: 1.0094033452608777; Bleu_1: 0.659; Bleu_2: 0.484; Bleu_3: 0.339; Bleu_4: 0.232; ; METEOR: 0.214; ROUGE_L: 0.488; CIDEr: 0.640

**Best and most Simple** (`v102_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs10`)
* my simple model - 1 epochs:
> ratio: 1.009220942322141; Bleu_1: 0.587; Bleu_2: 0.383; Bleu_3: 0.244; Bleu_4: 0.157; METEOR: 0.182; ROUGE_L: 0.424; CIDEr: 0.507
* my simple model - 2 epochs:
> ratio: 1.0005286538264253; Bleu_1: 0.616; Bleu_2: 0.416; Bleu_3: 0.275; Bleu_4: 0.183; METEOR: 0.192; ROUGE_L: 0.442; CIDEr: 0.574
* my simple model - 3 epochs:
> ratio: 1.014850954868193; Bleu_1: 0.637; Bleu_2: 0.449; Bleu_3: 0.306; Bleu_4: 0.209; METEOR: 0.211; ROUGE_L: 0.466; CIDEr: 0.667
* my simple model - 4 epochs:
> ratio: 1.0104244476961306; Bleu_1: 0.636
Bleu_2: 0.448; Bleu_3: 0.305; Bleu_4: 0.209; METEOR: 0.210; ROUGE_L: 0.464; CIDEr: 0.669
* my simple model - 5 epochs:
> ratio: 1.012540160924407; Bleu_1: 0.632; Bleu_2: 0.446; Bleu_3: 0.306; Bleu_4: 0.211; METEOR: 0.213; ROUGE_L: 0.466; CIDEr: 0.668


**Attention Before LSTM** (`v121_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs3`)
* my att first model - 1 epochs:
> ratio: 1.102866731062698; Bleu_1: 0.357; Bleu_2: 0.179; Bleu_3: 0.065; Bleu_4: 0.029; METEOR: 0.091; ROUGE_L: 0.285; CIDEr: 0.037
* my att first model - 2 epochs:
> ratio: 0.706783945336399; Bleu_1: 0.205; Bleu_2: 0.091; Bleu_3: 0.032; Bleu_4: 0.014; METEOR: 0.050; ROUGE_L: 0.174; CIDEr: 0.015

**Attention After LSTM** (`v120_paramsDecoder_withoutFlipTrans_batch_size10_vocabThr3_embedSize512_hiddenSize1024_totEpochs3`)
* my att aft model - 1 epochs:
> ratio: 0.33351825744295616; Bleu_1: 0.052; Bleu_2: 0.034; Bleu_3: 0.021; Bleu_4: 0.014; METEOR: 0.037; ROUGE_L: 0.101; CIDEr: 0.071
* my att aft model - 2 epochs:
> ratio: 0.21256804304037424; Bleu_1: 0.005; Bleu_2: 0.005; Bleu_3: 0.004; Bleu_4: 0.002; METEOR: 0.014; ROUGE_L: 0.034; CIDEr: 0.029


### Thoughts
My best model is clearly 1.0.2 (although simple), proved to be better than adding a multihead attention.
And performance wise is close to the most complicated one used to benchmark.
